drivers/power/max17851/max17851.c

```diff
@@ -0,0 +1,760 @@
+ * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include "max17851.h"
+#include "no_os_alloc.h"
+#include "no_os_error.h"
+
+/**
+ * @brief Read any UART related errors in case of failures communication wise.
+ * @param desc - Instance of UART
+ * @return Status byte that contains any UART errors that appeared.
+ */
+static uint32_t max17851_uart_get_errors(struct no_os_uart_desc *desc)
+{
+	struct max17851_desc *extra;
+	uint8_t reg_val;
+	int ret;
+
+	extra = desc->extra;
+	ret = max17851_reg_read(extra, MAX17851_STATUS_GEN_REG, &reg_val);
+	if (ret)
+		return ret;
+
+	return (int32_t)reg_val;
+}
+
+/**
+ * @brief Read data from UART device.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int32_t max17851_uart_read(struct no_os_uart_desc *desc, uint8_t *data,
+				  uint32_t bytes_number)
+{
+	struct max17851_desc *extra;
+
+	if (!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	if (bytes_number > 86)
+		return -EOPNOTSUPP;
+
+	extra = desc->extra;
+	return max17851_read_msg(extra, data, bytes_number);
+}
+
+/**
+ * @brief Write data to UART device.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+static int32_t max17851_uart_write(struct no_os_uart_desc *desc,
+				   const uint8_t *data,
+				   uint32_t bytes_number)
+{
+	struct max17851_desc *extra;
+
+	if (!desc || !data || !bytes_number)
+		return -EINVAL;
+
+	if (bytes_number > 86)
+		return -EOPNOTSUPP;
+
+	extra = desc->extra;
+	return max17851_write_msg(extra, data, bytes_number);
+}
+
+/**
+ * @brief Initialize the UART communication peripheral.
+ * @param desc - The UART descriptor.
+ * @param param - The structure that contains the UART parameters.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+static int32_t max17851_uart_init(struct no_os_uart_desc **desc,
+				  struct no_os_uart_init_param *param)
+{
+	struct max17851_desc *max17851_uart;
+	struct max17851_init_param *eparam;
+	struct no_os_uart_desc *descriptor;
+	int ret;
+
+	if (!param || !param->extra)
+		return -EINVAL;
+
+	descriptor = no_os_calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	eparam = param->extra;
+	ret = max17851_init(&max17851_uart, eparam);
+	if (ret)
+		goto error;
+
+	descriptor->extra = max17851_uart;
+	descriptor->device_id = 0;
+
+	switch (eparam->baud_rate) {
+	case MAX17851_UART_BAUD_500K:
+		descriptor->baud_rate = 500000;
+
+		break;
+	case MAX17851_UART_BAUD_1M:
+		descriptor->baud_rate = 1000000;
+
+		break;
+	case MAX17851_UART_BAUD_2M:
+		descriptor->baud_rate = 2000000;
+
+		break;
+	default:
+		return -EINVAL;
+	}
```

Comment:
This looks a bit strange. Shouldn't we set the baudrate based on what is given in the initialization parameter? Also you're not calling `no_os_free()` and `max17851_init()` in the default case.